packages/gatsby/src/redux/nodes.js

```diff
@@ -154,3 +154,125 @@ const addResolvedNodes = (typeName, arr) => {
 }
 
 exports.addResolvedNodes = addResolvedNodes
+
+/**
+ * Given a ("flat") filter path leading up to "eq", a set of node types, and a
+ * cache, create a cache that for each resulting value of the filter contains
+ * all the Nodes in a Set (or, if the property is `id`, just the Nodes).
+ * This cache is used for applying the filter and is a massive improvement over
+ * looping over all the nodes, when the number of pages (/nodes) scale up.
+ *
+ * @param {Array<string>} chain
+ * @param {Array<string>} nodeTypeNames
+ * @param {undefined | Map<string, Map<string | number | boolean, Node>>} typedKeyValueIndexes
+ *   This object lives in query/query-runner.js and is passed down runQuery
+ * @returns {undefined}
+ */
+const ensureIndexByTypedChain = (
+  chain,
+  nodeTypeNames,
+  typedKeyValueIndexes
+) => {
+  const chained = chain.join(`+`)
+
+  const nodeTypeNamePrefix = nodeTypeNames.join(`,`) + `/`
+  // The format of the typedKey is `type,type/path+to+eqobj`
+  const typedKey = nodeTypeNamePrefix + chained
+
+  let byKeyValue = typedKeyValueIndexes.get(typedKey)
```

Comment:
DRY nitpick (to avoid different `typedKey` generation):

until this line it is 95% of `getNodesByTypedChain` ...

Reply:
You're right, that's a bit of DRY. The reason is that `chained` is referenced again later to check whether the path is directly filtering for `id`. I'm open to suggestions to doing that better. Would you prefer a redundant `chain.join('+')` for this check and then abstraction for the `typedKey` bit?